stan/math/rev/mat/fun/sqrt_spd.hpp

```diff
@@ -0,0 +1,62 @@
+#ifndef STAN_MATH_REV_MAT_FUN_SQRT_SPD_HPP
+#define STAN_MATH_REV_MAT_FUN_SQRT_SPD_HPP
+
+#include <cstddef>
+#include <stan/math/prim/mat/fun/sqrt_spd.hpp>
+#include <stan/math/rev/mat/functor/adj_jac_apply.hpp>
+#include <tuple>
+#include <unsupported/Eigen/KroneckerProduct>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace {
+class sqrt_spd_op {
+  int K_;
+  double *y_; // Holds the output, which has K_^2 elements
+
+public:
+  sqrt_spd_op() : K_(0), y_(nullptr) {}
+
+  template <std::size_t size>
+  Eigen::MatrixXd operator()(const std::array<bool, size> &needs_adj,
+                             const Eigen::MatrixXd &m) {
+    K_ = m.rows();
+    Eigen::MatrixXd sqrt_m = sqrt_spd(m);
+    int KK = K_ * K_;
+    y_ = ChainableStack::instance().memalloc_.alloc_array<double>(KK);
+    for (int k = 0; k < KK; ++k)
+      y_[k] = sqrt_m.coeff(k);
+    return sqrt_m;
+  }
+
+  //  https://math.stackexchange.com/questions/540361/
+  //  derivative-or-differential-of-symmetric-square-root-of-a-matrix
+
+  template <std::size_t size>
+  std::tuple<Eigen::VectorXd>
+  multiply_adjoint_jacobian(const std::array<bool, size> &needs_adj,
+                            const Eigen::VectorXd &adj) const {
+    using Eigen::kroneckerProduct;
+    using Eigen::MatrixXd;
+    Eigen::Map<MatrixXd> sqrt_m(y_, K_, K_);
+    return std::make_tuple(
+        (kroneckerProduct(sqrt_m, MatrixXd::Identity(K_, K_)) +
+         kroneckerProduct(MatrixXd::Identity(K_, K_), sqrt_m))
+            // the above is symmetric so can skip the transpose
+            .ldlt()
+            .solve(adj)
+            .eval());
+  }
+};
+} // namespace
+
+inline Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>
+sqrt_spd(const Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic> &m) {
+  return adj_jac_apply<sqrt_spd_op>(m);
```

Comment:
Add symmetric + square checks

Reply:
Thanks for looking into this. My impression was that checking symmetry in `prim/mat/fun/sqrt_spd.hpp` would be enough if the rev version is calling the prim version. Is there some scenario in which `adj_jac_apply` would be calculating something on a possibly asymmetric matrix before the prim version of `sqrt_spd` was called?

Reply:
Oh yeah, my bad, this looks good.